airflow/providers/amazon/aws/operators/s3.py

```diff
@@ -503,23 +509,37 @@ def __init__(
         self.bucket = bucket
         self.keys = keys
         self.prefix = prefix
+        self.from_datetime = from_datetime
+        self.to_datetime = to_datetime
         self.aws_conn_id = aws_conn_id
         self.verify = verify
 
         self._keys: str | list[str] = ""
 
-        if not exactly_one(prefix is None, keys is None):
-            raise AirflowException("Either keys or prefix should be set.")
+        if not exactly_one(keys is None, all([prefix is None, from_datetime is None, to_datetime is None])):
+            raise AirflowException(
+                "Either keys or at least one of prefix, from_datetime, to_datetime should be set."
```

Comment:
That message does not match what the code is enforcing.  The code says either keys or all of the others, no? Isn't this saying either keys can be empty or all the rest are?  Maybe I'm just getting confused by the double negative here.

What is you change it to a positive check like:


```suggestion
        if not exactly_one(keys, at_least_one(prefix, from_datetime, to_datetime])):
```

Reply:
I was playing with that in ipython and looking at the unit tests, and apparently an empty list and an empty string are considered valid values?   That feels wrong to me.  If one must be provided, then surely providing an empty value shouldn't count?   That complicates things and I guess explains why this statement is convoluted.

My initial statement stands though, I'm pretty sure this is not testing what your error message says it is.

Reply:
I agree that it will be easier to understand without the negative sections.
do you suggest to add the exactlyone function to airflow helpers?
The statement checks if keys are given or at least one of the other params were given.
it could be only the keys or any combination that includes one of the other params

about empty list or empty string they are both valid options for keys, maybe someone has this operator in a dag that gets dynamically which keys to delete but for a specific run he didn't have any keys to delete so the value of keys will be the empty list and the action that will be performed is to not delete any key

Reply:
there are specific tests that check what happens for empty list or string in keys value

Reply:
> there are specific tests that check what happens for empty list or string in keys value

Yeah I saw those, I just disagree with how they were handled, but that's way out of scope for this PR.  If an empty value is acceptable, then it's clearly NOT a required value.  But either way, not something to fix here. Sorry about that.


> do you suggest to add the exactlyone function to airflow helpers?

If you feel like doing it, I'd support that, but I won't require it.   This statementis correct as written but really isn't intuitive at all, and that's usually when I start looking for a helper to clarify.  We already have `exactly_one` and `at_most_one`, it makes sense ti add `at_least_one`.


Feel free to resolve this convo.